File: blobs/howto/dotnet/dotnet-v12/ACL_DataLake.cs
//----------------------------------------------------------------------------------
// Microsoft Developer & Platform Evangelism
//
// Copyright (c) Microsoft Corporation. All rights reserved.
//
// THIS CODE AND INFORMATION ARE PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND, 
// EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED WARRANTIES 
// OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
//----------------------------------------------------------------------------------
// The example companies, organizations, products, domain names,
// e-mail addresses, logos, people, places, and events depicted
// herein are fictitious.  No association with any real company,
// organization, product, domain name, email address, logo, person,
// places, or events is intended or should be inferred.
//----------------------------------------------------------------------------------

using Azure.Storage.Files.DataLake;
using System;
using System.Threading.Tasks;
using Azure.Storage.Files.DataLake.Models;
using System.Collections.Generic;

namespace dotnet_v12
{
    public class ACL_DataLake
    {

        DataLakeServiceClient dataLakeServiceClient;

        #region Get file system

        // ---------------------------------------------------------
        // Get a fileSystem
        //----------------------------------------------------------

        // <Snippet_GetFileSystem>
        public DataLakeFileSystemClient GetFileSystem
            (DataLakeServiceClient serviceClient, string fileSystemName)
        {
            DataLakeFileSystemClient fileSystemClient =
                serviceClient.GetFileSystemClient(fileSystemName);

            return fileSystemClient;
        }
        // </Snippet_GetFileSystem>

        #endregion

        #region Get and set directory ACLs

        // ---------------------------------------------------------
        // Get and set directory-level ACLs
        //----------------------------------------------------------

        // <Snippet_ACLDirectory>
        public async Task ManageDirectoryACLs(DataLakeFileSystemClient fileSystemClient)
        {
            DataLakeDirectoryClient directoryClient =
              fileSystemClient.GetDirectoryClient("");

            PathAccessControl directoryAccessControl =
                await directoryClient.GetAccessControlAsync();

            foreach (var item in directoryAccessControl.AccessControlList)
            {
                Console.WriteLine(item.ToString());
            }

            IList<PathAccessControlItem> accessControlList
                = PathAccessControlExtensions.ParseAccessControlList
                ("user::rwx,group::r-x,other::rw-");

            directoryClient.SetAccessControlList(accessControlList);

        }
        // </Snippet_ACLDirectory>

        #endregion

        #region Get and set file ACLs

        // ---------------------------------------------------------
        // Get and set ACLs on files
        //----------------------------------------------------------

        // <Snippet_FileACL>
        public async Task ManageFileACLs(DataLakeFileSystemClient fileSystemClient)
        {
            DataLakeDirectoryClient directoryClient =
                fileSystemClient.GetDirectoryClient("my-directory");

            DataLakeFileClient fileClient =
                directoryClient.GetFileClient("hello.txt");

            PathAccessControl FileAccessControl =
                await fileClient.GetAccessControlAsync();

            foreach (var item in FileAccessControl.AccessControlList)
            {
                Console.WriteLine(item.ToString());
            }

            IList<PathAccessControlItem> accessControlList
                = PathAccessControlExtensions.ParseAccessControlList
                ("user::rwx,group::r-x,other::rw-");

            fileClient.SetAccessControlList(accessControlList);
        }
        // </Snippet_FileACL>

        #endregion

        #region Update ACL
        // ---------------------------------------------------------
        // Update directory-level ACLs
        //----------------------------------------------------------

        // <Snippet_UpdateACL>
        public async Task UpdateDirectoryACLs(DataLakeFileSystemClient fileSystemClient)
        {
            DataLakeDirectoryClient directoryClient =
              fileSystemClient.GetDirectoryClient("");

            PathAccessControl directoryAccessControl =
                await directoryClient.GetAccessControlAsync();

            List<PathAccessControlItem> accessControlListUpdate 
                = (List<PathAccessControlItem>)directoryAccessControl.AccessControlList;

            int index = -1;

            foreach (var item in accessControlListUpdate)
            {
                if (item.AccessControlType == AccessControlType.Other)
                {
                    index = accessControlListUpdate.IndexOf(item);
                    break;
                }
            }

            if (index > -1)
            {
                accessControlListUpdate[index] = new PathAccessControlItem(AccessControlType.Other,
                RolePermissions.Read |
                RolePermissions.Execute);

                directoryClient.SetAccessControlList(accessControlListUpdate);
            }

           }

            // </Snippet_UpdateACL>

            #endregion

        #region Set ACLs recursively

            // <Snippet_SetACLRecursively>
            public async Task SetACLRecursively(DataLakeServiceClient serviceClient, bool isDefaultScope)
        {
            DataLakeDirectoryClient directoryClient =
                serviceClient.GetFileSystemClient("my-container").
                    GetDirectoryClient("my-parent-directory");

            List<PathAccessControlItem> accessControlList =
                new List<PathAccessControlItem>()
            {
        new PathAccessControlItem(AccessControlType.User,
            RolePermissions.Read |
            RolePermissions.Write |
            RolePermissions.Execute, isDefaultScope),

        new PathAccessControlItem(AccessControlType.Group,
            RolePermissions.Read |
            RolePermissions.Execute, isDefaultScope),

        new PathAccessControlItem(AccessControlType.Other,
            RolePermissions.None, isDefaultScope),

        new PathAccessControlItem(AccessControlType.User,
            RolePermissions.Read |
            RolePermissions.Execute, isDefaultScope,
            entityId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"),
            };

            await directoryClient.SetAccessControlRecursiveAsync
                (accessControlList, null);
        }
        // </Snippet_SetACLRecursively>

        #endregion

        #region Update ACLs recursively

        // <Snippet_UpdateACLsRecursively>
        public async Task UpdateACLsRecursively(DataLakeServiceClient serviceClient, bool isDefaultScope)
        {
            DataLakeDirectoryClient directoryClient =
                serviceClient.GetFileSystemClient("my-container").
                GetDirectoryClient("my-parent-directory");

            List<PathAccessControlItem> accessControlListUpdate =
                new List<PathAccessControlItem>()
            {
        new PathAccessControlItem(AccessControlType.User,
            RolePermissions.Read |
            RolePermissions.Write |
            RolePermissions.Execute, isDefaultScope,
            entityId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"),
            };

            await directoryClient.UpdateAccessControlRecursiveAsync
                (accessControlListUpdate, null);

        }
        // </Snippet_UpdateACLsRecursively>

        #endregion

        #region Remove ACL entry
        // ---------------------------------------------------------
        // Remove directory-level ACL entry
        //----------------------------------------------------------

        // <Snippet_RemoveACLEntry>
        public async Task RemoveDirectoryACLEntry
            (DataLakeFileSystemClient fileSystemClient)
        {
            DataLakeDirectoryClient directoryClient =
              fileSystemClient.GetDirectoryClient("");

            PathAccessControl directoryAccessControl =
                await directoryClient.GetAccessControlAsync();

            List<PathAccessControlItem> accessControlListUpdate
                = (List<PathAccessControlItem>)directoryAccessControl.AccessControlList;

            PathAccessControlItem entryToRemove = null;

            foreach (var item in accessControlListUpdate)
            {
                if (item.EntityId == "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx")
                {
                    entryToRemove = item;
                    break;
                }
            }

            if (entryToRemove != null)
            {
                accessControlListUpdate.Remove(entryToRemove);
                directoryClient.SetAccessControlList(accessControlListUpdate);
            }

        }
        // </Snippet_RemoveACLEntry>

        #endregion

        #region Remove ACLs recursively

        // <Snippet_RemoveACLRecursively>
        public async Task RemoveACLsRecursively(DataLakeServiceClient serviceClient, bool isDefaultScope)
        {
            DataLakeDirectoryClient directoryClient =
                serviceClient.GetFileSystemClient("my-container").
                    GetDirectoryClient("my-parent-directory");

            List<RemovePathAccessControlItem> accessControlListForRemoval =
                new List<RemovePathAccessControlItem>()
                {
            new RemovePathAccessControlItem(AccessControlType.User, isDefaultScope,
            entityId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"),
                };

            await directoryClient.RemoveAccessControlRecursiveAsync
                (accessControlListForRemoval, null);

        }
        // </Snippet_RemoveACLRecursively>

        #endregion

        #region Resume with event token

        // <Snippet_ResumeContinuationToken>
        public async Task<string> ResumeAsync(DataLakeServiceClient serviceClient,
            DataLakeDirectoryClient directoryClient,
            List<PathAccessControlItem> accessControlList,
            string continuationToken)
        {
            try
            {
                var accessControlChangeResult =
                    await directoryClient.SetAccessControlRecursiveAsync(
                        accessControlList, continuationToken: continuationToken, null);

                if (accessControlChangeResult.Value.Counters.FailedChangesCount > 0)
                {
                    continuationToken =
                        accessControlChangeResult.Value.ContinuationToken;
                }

                return continuationToken;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
                return continuationToken;
            }

        }
        // </Snippet_ResumeContinuationToken>

        #endregion

        #region Continue on failure

        // <Snippet_ContinueOnFailure>
        public async Task ContinueOnFailureAsync(DataLakeServiceClient serviceClient,
            DataLakeDirectoryClient directoryClient,
            List<PathAccessControlItem> accessControlList)
        {
            var accessControlChangeResult =
                await directoryClient.SetAccessControlRecursiveAsync(
                    accessControlList, null, new AccessControlChangeOptions()
                    { ContinueOnFailure = true });

            var counters = accessControlChangeResult.Value.Counters;

            Console.WriteLine("Number of directories changed: " +
                counters.ChangedDirectoriesCount.ToString());

            Console.WriteLine("Number of files changed: " +
                counters.ChangedFilesCount.ToString());

            Console.WriteLine("Number of failures: " +
                counters.FailedChangesCount.ToString());
        }
        // </Snippet_ContinueOnFailure>

        #endregion

        //-------------------------------------------------
        // Security menu (Can call asynchronous and synchronous methods)
        //-------------------------------------------------

        public async Task<bool> MenuAsync()
        {
            Console.Clear();
            Console.WriteLine("Choose a security scenario:");
            Console.WriteLine("1) Set a directory ACL");
            Console.WriteLine("2) Set a file ACL");
            Console.WriteLine("3) Update an ACL");
            Console.WriteLine("4) Set ACLs recursively");
            Console.WriteLine("5) Update ACLs recursively");
            Console.WriteLine("6) Remove an ACL entry");
            Console.WriteLine("7) Remove ACLs recursively");
            Console.WriteLine("8) Resume after failure with token");
            Console.WriteLine("9) Continue past failure");

            Console.WriteLine("X) Exit to main menu");
            Console.Write("\r\nSelect an option: ");

            // Uncomment if you want to test shared key auth.
            DataLakeServiceClient dataLakeServiceClient = 
                Authorize_DataLake.GetDataLakeServiceClient(Constants.storageAccountName, Constants.accountKey);

            // Uncomment if you want to test AAD auth.
            //DataLakeServiceClient dataLakeServiceClient = 
            //    Authorize_DataLake.GetDataLakeServiceClient(Constants.storageAccountName);

            // Get file system client

            DataLakeFileSystemClient fileSystemClient =
                GetFileSystem(dataLakeServiceClient, Constants.containerName);

            switch (Console.ReadLine())
            {
                case "1":

                    await ManageDirectoryACLs(fileSystemClient);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "2":

                    await ManageFileACLs(fileSystemClient);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "3":

                    await UpdateDirectoryACLs(fileSystemClient);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "4":

                    await SetACLRecursively(dataLakeServiceClient, false);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "5":

                    await UpdateACLsRecursively(dataLakeServiceClient, false);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "6":

                    await RemoveDirectoryACLEntry(fileSystemClient);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "7":

                    await RemoveACLsRecursively(dataLakeServiceClient, false);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "8":

                    DataLakeDirectoryClient directoryClient =
                        dataLakeServiceClient.GetFileSystemClient("my-container").
                        GetDirectoryClient("my-parent-directory");

                    directoryClient =
                         dataLakeServiceClient.GetFileSystemClient("my-container").
                         GetDirectoryClient("my-parent-directory");

                    List<PathAccessControlItem> accessControlList = new List<PathAccessControlItem>()
                    {
                        new PathAccessControlItem(AccessControlType.User,
                            RolePermissions.Read |
                            RolePermissions.Write |
                            RolePermissions.Execute),

                        new PathAccessControlItem(AccessControlType.Group,
                            RolePermissions.Read |
                            RolePermissions.Execute),

                        new PathAccessControlItem(AccessControlType.Other,
                            RolePermissions.None),

                        new PathAccessControlItem(AccessControlType.User, RolePermissions.Read |
                            RolePermissions.Write | RolePermissions.Execute,
                            entityId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx"),

                    };

                    await ResumeAsync(dataLakeServiceClient, directoryClient, accessControlList, null);


                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "9":

                    DataLakeDirectoryClient directoryClient2 =
                        dataLakeServiceClient.GetFileSystemClient("my-container").
                        GetDirectoryClient("my-parent-directory");

                    List<PathAccessControlItem> accessControlList2 = new List<PathAccessControlItem>()
                     {
                         new PathAccessControlItem(AccessControlType.User,
                             RolePermissions.Read |
                             RolePermissions.Write |
                             RolePermissions.Execute),

                         new PathAccessControlItem(AccessControlType.Group,
                             RolePermissions.Read |
                             RolePermissions.Execute),

                         new PathAccessControlItem(AccessControlType.Other,
                             RolePermissions.None),

                         new PathAccessControlItem(AccessControlType.User, RolePermissions.Read |
                             RolePermissions.Write | RolePermissions.Execute,
                             entityId: "xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxx"),

                     };

                    await ContinueOnFailureAsync(dataLakeServiceClient, directoryClient2, accessControlList2);

                    Console.WriteLine("Press enter to continue");
                    Console.ReadLine();
                    return true;

                case "x":
                case "X":

                    return false;

                default:

                    return true;
            }
        }
    }
        
    }

File: blobs/howto/Java/Java-v12/src/main/java/com/datalake/manage/ACL_DataLake.java
package com.datalake.manage;

import java.io.BufferedReader;

import java.io.InputStreamReader;

import java.security.InvalidKeyException;
import java.net.URISyntaxException;
import java.util.ArrayList;
import java.util.List;

import com.azure.core.http.rest.Response;
import com.azure.storage.file.datalake.DataLakeDirectoryClient;
import com.azure.storage.file.datalake.DataLakeFileClient;
import com.azure.storage.file.datalake.DataLakeFileSystemClient;
import com.azure.storage.file.datalake.DataLakeServiceClient;
import com.azure.storage.file.datalake.models.AccessControlChangeCounters;
import com.azure.storage.file.datalake.models.AccessControlChangeResult;
import com.azure.storage.file.datalake.models.AccessControlType;
import com.azure.storage.file.datalake.models.PathAccessControl;
import com.azure.storage.file.datalake.models.PathAccessControlEntry;
import com.azure.storage.file.datalake.models.PathPermissions;
import com.azure.storage.file.datalake.models.PathRemoveAccessControlEntry;
import com.azure.storage.file.datalake.models.RolePermissions;
import com.azure.storage.file.datalake.options.PathSetAccessControlRecursiveOptions;

public class ACL_DataLake {
    
    // ----------------------------------------------------------
    // Get a file system
    // ----------------------------------------------------------
    
    // <Snippet_GetFileSystem>
    public DataLakeFileSystemClient GetFileSystem
       (DataLakeServiceClient serviceClient, String fileSystemName){
        
        DataLakeFileSystemClient fileSystemClient =
            serviceClient.getFileSystemClient(fileSystemName);

        return fileSystemClient;
    }
    // </Snippet_GetFileSystem>
    
    // ----------------------------------------------------------
    // Get and set directory-level permissions
    // ----------------------------------------------------------

    // <Snippet_ManageDirectoryACLs>
    public void ManageDirectoryACLs(DataLakeFileSystemClient fileSystemClient){

        DataLakeDirectoryClient directoryClient =
          fileSystemClient.getDirectoryClient("");

        PathAccessControl directoryAccessControl =
            directoryClient.getAccessControl();

        List<PathAccessControlEntry> pathPermissions = directoryAccessControl.getAccessControlList();
       
        System.out.println(PathAccessControlEntry.serializeList(pathPermissions));
             
        RolePermissions groupPermission = new RolePermissions();
        groupPermission.setExecutePermission(true).setReadPermission(true);
  
        RolePermissions ownerPermission = new RolePermissions();
        ownerPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);
  
        RolePermissions otherPermission = new RolePermissions();
        otherPermission.setReadPermission(true);
  
        PathPermissions permissions = new PathPermissions();
  
        permissions.setGroup(groupPermission);
        permissions.setOwner(ownerPermission);
        permissions.setOther(otherPermission);

        directoryClient.setPermissions(permissions, null, null);

        pathPermissions = directoryClient.getAccessControl().getAccessControlList();
     
        System.out.println(PathAccessControlEntry.serializeList(pathPermissions));

    }
    // </Snippet_ManageDirectoryACLs>

    // ----------------------------------------------------------
    // Get and set file-level permissions
    // ----------------------------------------------------------

    // <Snippet_ManageFileACLs>
    public void ManageFileACLs(DataLakeFileSystemClient fileSystemClient){

        DataLakeDirectoryClient directoryClient =
          fileSystemClient.getDirectoryClient("my-directory");

        DataLakeFileClient fileClient = 
          directoryClient.getFileClient("uploaded-file.txt");

        PathAccessControl fileAccessControl =
            fileClient.getAccessControl();

      List<PathAccessControlEntry> pathPermissions = fileAccessControl.getAccessControlList();
     
      System.out.println(PathAccessControlEntry.serializeList(pathPermissions));
           
      RolePermissions groupPermission = new RolePermissions();
      groupPermission.setExecutePermission(true).setReadPermission(true);

      RolePermissions ownerPermission = new RolePermissions();
      ownerPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);

      RolePermissions otherPermission = new RolePermissions();
      otherPermission.setReadPermission(true);

      PathPermissions permissions = new PathPermissions();

      permissions.setGroup(groupPermission);
      permissions.setOwner(ownerPermission);
      permissions.setOther(otherPermission);

      fileClient.setPermissions(permissions, null, null);

      pathPermissions = fileClient.getAccessControl().getAccessControlList();
   
      System.out.println(PathAccessControlEntry.serializeList(pathPermissions));

    }
    // </Snippet_ManageFileACLs>

    //-------------------------------------------------
    // Set ACLs recursively
    //-------------------------------------------------

    // <Snippet_SetACLRecursively>
    public void SetACLRecursively(DataLakeFileSystemClient fileSystemClient, Boolean isDefaultScope){
        
        DataLakeDirectoryClient directoryClient =
            fileSystemClient.getDirectoryClient("my-parent-directory");

        List<PathAccessControlEntry> pathAccessControlEntries = 
            new ArrayList<PathAccessControlEntry>();

        // Create owner entry.
        PathAccessControlEntry ownerEntry = new PathAccessControlEntry();

        RolePermissions ownerPermission = new RolePermissions();
        ownerPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);

        ownerEntry.setDefaultScope(isDefaultScope);
        ownerEntry.setAccessControlType(AccessControlType.USER);
        ownerEntry.setPermissions(ownerPermission);

        pathAccessControlEntries.add(ownerEntry);

        // Create group entry.
        PathAccessControlEntry groupEntry = new PathAccessControlEntry();

        RolePermissions groupPermission = new RolePermissions();
        groupPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(false);

        groupEntry.setDefaultScope(isDefaultScope);
        groupEntry.setAccessControlType(AccessControlType.GROUP);
        groupEntry.setPermissions(groupPermission);

        pathAccessControlEntries.add(groupEntry);

        // Create other entry.
        PathAccessControlEntry otherEntry = new PathAccessControlEntry();

        RolePermissions otherPermission = new RolePermissions();
        otherPermission.setExecutePermission(false).setReadPermission(false).setWritePermission(false);

        otherEntry.setDefaultScope(isDefaultScope);
        otherEntry.setAccessControlType(AccessControlType.OTHER);
        otherEntry.setPermissions(otherPermission);

        pathAccessControlEntries.add(otherEntry);

        // Create named user entry.
        PathAccessControlEntry userEntry = new PathAccessControlEntry();

        RolePermissions userPermission = new RolePermissions();
        userPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(false);

        userEntry.setDefaultScope(isDefaultScope);
        userEntry.setAccessControlType(AccessControlType.USER);
        userEntry.setEntityId("xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx");
        userEntry.setPermissions(userPermission);    
        
        pathAccessControlEntries.add(userEntry);
        
        directoryClient.setAccessControlRecursive(pathAccessControlEntries);        

    }
    // </Snippet_SetACLRecursively>

    //-------------------------------------------------
    // Update an ACL
    //-------------------------------------------------

    // <Snippet_UpdateACL>
    public void UpdateACL(DataLakeFileSystemClient fileSystemClient, Boolean isDefaultScope){

        DataLakeDirectoryClient directoryClient =
        fileSystemClient.getDirectoryClient("my-parent-directory");

        List<PathAccessControlEntry> pathAccessControlEntries = 
            directoryClient.getAccessControl().getAccessControlList();

        int index = -1;

        for (PathAccessControlEntry pathAccessControlEntry : pathAccessControlEntries){
            
            if (pathAccessControlEntry.getAccessControlType() == AccessControlType.OTHER){
                index = pathAccessControlEntries.indexOf(pathAccessControlEntry);
                break;
            }
        }

        if (index > -1){

        PathAccessControlEntry userEntry = new PathAccessControlEntry();

        RolePermissions userPermission = new RolePermissions();
        userPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);

        userEntry.setDefaultScope(isDefaultScope);
        userEntry.setAccessControlType(AccessControlType.OTHER);
        userEntry.setPermissions(userPermission);  
        
        pathAccessControlEntries.set(index, userEntry);
        }
 
        directoryClient.setAccessControlList(pathAccessControlEntries, 
            directoryClient.getAccessControl().getGroup(), 
            directoryClient.getAccessControl().getOwner());
    
    }
    // </Snippet_UpdateACL>


    //-------------------------------------------------
    // Update ACLs recursively
    //-------------------------------------------------

    // <Snippet_UpdateACLRecursively>
    public void UpdateACLRecursively(DataLakeFileSystemClient fileSystemClient, Boolean isDefaultScope){

        DataLakeDirectoryClient directoryClient =
        fileSystemClient.getDirectoryClient("my-parent-directory");

        List<PathAccessControlEntry> pathAccessControlEntries = 
            new ArrayList<PathAccessControlEntry>();

        // Create named user entry.
        PathAccessControlEntry userEntry = new PathAccessControlEntry();

        RolePermissions userPermission = new RolePermissions();
        userPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);

        userEntry.setDefaultScope(isDefaultScope);
        userEntry.setAccessControlType(AccessControlType.USER);
        userEntry.setEntityId("xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx");
        userEntry.setPermissions(userPermission);    
        
        pathAccessControlEntries.add(userEntry);
        
        directoryClient.updateAccessControlRecursive(pathAccessControlEntries);        
    
    }
    // </Snippet_UpdateACLRecursively>

    //-------------------------------------------------
    // Remove an ACL entry
    //-------------------------------------------------

    // <Snippet_RemoveACLEntry>
    public void RemoveACLEntry(DataLakeFileSystemClient fileSystemClient, Boolean isDefaultScope){

        DataLakeDirectoryClient directoryClient =
        fileSystemClient.getDirectoryClient("my-parent-directory");

        List<PathAccessControlEntry> pathAccessControlEntries = 
            directoryClient.getAccessControl().getAccessControlList();

        PathAccessControlEntry entryToRemove = null;

        for (PathAccessControlEntry pathAccessControlEntry : pathAccessControlEntries){
            
            if (pathAccessControlEntry.getEntityId() != null){

                if (pathAccessControlEntry.getEntityId().equals("xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx")){
                    entryToRemove = pathAccessControlEntry;
                    break;
                }
            }

        }

        if (entryToRemove != null){
            
            pathAccessControlEntries.remove(entryToRemove);

            directoryClient.setAccessControlList(pathAccessControlEntries, 
            directoryClient.getAccessControl().getGroup(), 
            directoryClient.getAccessControl().getOwner());   
        }
   
    
    }
    // </Snippet_RemoveACLEntry>

    //-------------------------------------------------
    // Remove ACLs recursively
    //-------------------------------------------------

    // <Snippet_RemoveACLRecursively>
    public void RemoveACLEntryRecursively(DataLakeFileSystemClient fileSystemClient, Boolean isDefaultScope){

        DataLakeDirectoryClient directoryClient =
        fileSystemClient.getDirectoryClient("my-parent-directory");
  
        List<PathRemoveAccessControlEntry> pathRemoveAccessControlEntries = 
            new ArrayList<PathRemoveAccessControlEntry>();

        // Create named user entry.
        PathRemoveAccessControlEntry userEntry = new PathRemoveAccessControlEntry();

        RolePermissions userPermission = new RolePermissions();
        userPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);

        userEntry.setDefaultScope(isDefaultScope);
        userEntry.setAccessControlType(AccessControlType.USER);
        userEntry.setEntityId("xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx"); 
        
        pathRemoveAccessControlEntries.add(userEntry);
        
        directoryClient.removeAccessControlRecursive(pathRemoveAccessControlEntries);      
    
    }
    // </Snippet_RemoveACLRecursively>

    //--------------------------------------------------
    // Use continuation token
    //--------------------------------------------------

    // <Snippet_ResumeSetACLRecursively>
    public String ResumeSetACLRecursively(DataLakeFileSystemClient fileSystemClient,
        DataLakeDirectoryClient directoryClient,
        List<PathAccessControlEntry> accessControlList, 
        String continuationToken){

        try{
            PathSetAccessControlRecursiveOptions options = new PathSetAccessControlRecursiveOptions(accessControlList);
            
            options.setContinuationToken(continuationToken);
        
           Response<AccessControlChangeResult> accessControlChangeResult =  
              directoryClient.setAccessControlRecursiveWithResponse(options, null, null);

           if (accessControlChangeResult.getValue().getCounters().getFailedChangesCount() > 0)
           {
              continuationToken =
                  accessControlChangeResult.getValue().getContinuationToken();
           }
        
           return continuationToken;

        }
        catch(Exception ex){
        
            System.out.println(ex.toString());
            return continuationToken;
        }


    }
    // </Snippet_ResumeSetACLRecursively>

    //--------------------------------------------------
    // Continue on failure
    //--------------------------------------------------

    // <Snippet_ContinueOnFailure>
    public void ContinueOnFailure(DataLakeFileSystemClient fileSystemClient,
        DataLakeDirectoryClient directoryClient,
        List<PathAccessControlEntry> accessControlList){
        
        PathSetAccessControlRecursiveOptions options = 
           new PathSetAccessControlRecursiveOptions(accessControlList);
            
        options.setContinueOnFailure(true);
        
        Response<AccessControlChangeResult> accessControlChangeResult =  
            directoryClient.setAccessControlRecursiveWithResponse(options, null, null);

        AccessControlChangeCounters counters = accessControlChangeResult.getValue().getCounters();

        System.out.println("Number of directories changes: " + 
            counters.getChangedDirectoriesCount());

        System.out.println("Number of files changed: " + 
            counters.getChangedDirectoriesCount());

        System.out.println("Number of failures: " + 
            counters.getChangedDirectoriesCount());
    }
    // </Snippet_ContinueOnFailure>

    // ----------------------------------------------------------
    // Driver menu
    // ----------------------------------------------------------

    public void ShowMenu() throws java.lang.Exception, URISyntaxException, InvalidKeyException{
    
        try {
            DataLakeServiceClient dataLakeServiceClient = Authorize_DataLake.GetDataLakeServiceClient
            (Constants.storageAccountName, Constants.accountKey);

            // Uncomment if you want to test AD Authorization.
            // DataLakeServiceClient dataLakeServiceClient = Authorize_DataLake.GetDataLakeServiceClient
            //    (accountName, clientID, clientSecret, tenantID);

            DataLakeFileSystemClient fileSystemClient = GetFileSystem(dataLakeServiceClient, Constants.containerName);

            // Listening for commands from the console
            System.out.print("\033[H\033[2J");  
            System.out.flush();

            System.out.println("Enter a command");

            System.out.println("(1) Manage directory ACL | (2) Manage File ACLs | (3) Set ACL recursive | " +
                            "(4) Update an ACL | (5) Update ACL recursive | | (6) Remove an ACL | (7) Remove ACL recursive " +
                            "|(8) Resume ACL recursive | (9) Exit");

            BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));

            while (true) {

                System.out.println("# Enter a command : ");
                String input = reader.readLine();

                switch(input){

                    case "1":
                        ManageDirectoryACLs(fileSystemClient);
                        break;
                    case "2":
                        ManageFileACLs(fileSystemClient);
                    break;
                    case "3":
                        SetACLRecursively(fileSystemClient, false);
                    break;
                    case "4":
                        UpdateACL(fileSystemClient, false);
                    break;
                    case "5":
                        UpdateACLRecursively(fileSystemClient, false);
                    break;
                    case "6":
                        RemoveACLEntry(fileSystemClient, false);
                    break;
                    case "7":
                        RemoveACLEntryRecursively(fileSystemClient, false);
                    break;
                    case "8":
                        DataLakeDirectoryClient directoryClient =
                        fileSystemClient.getDirectoryClient("my-parent-directory");

                        List<PathAccessControlEntry> pathAccessControlEntries = 
                            new ArrayList<PathAccessControlEntry>();

                        // Create owner entry.
                        PathAccessControlEntry ownerEntry = new PathAccessControlEntry();

                        RolePermissions ownerPermission = new RolePermissions();
                        ownerPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(true);

                        ownerEntry.setAccessControlType(AccessControlType.USER);
                        ownerEntry.setPermissions(ownerPermission);

                        pathAccessControlEntries.add(ownerEntry);

                        // Create group entry.
                        PathAccessControlEntry groupEntry = new PathAccessControlEntry();

                        RolePermissions groupPermission = new RolePermissions();
                        groupPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(false);

                        groupEntry.setAccessControlType(AccessControlType.GROUP);
                        groupEntry.setPermissions(groupPermission);

                        pathAccessControlEntries.add(groupEntry);

                        // Create other entry.
                        PathAccessControlEntry otherEntry = new PathAccessControlEntry();

                        RolePermissions otherPermission = new RolePermissions();
                        otherPermission.setExecutePermission(false).setReadPermission(false).setWritePermission(false);

                        otherEntry.setAccessControlType(AccessControlType.OTHER);
                        otherEntry.setPermissions(otherPermission);

                        pathAccessControlEntries.add(otherEntry);

                        // Create named user entry.
                        PathAccessControlEntry userEntry = new PathAccessControlEntry();

                        RolePermissions userPermission = new RolePermissions();
                        userPermission.setExecutePermission(true).setReadPermission(true).setWritePermission(false);

                        userEntry.setAccessControlType(AccessControlType.USER);
                        userEntry.setEntityId("4a9028cf-f779-4032-b09d-970ebe3db258");
                        userEntry.setPermissions(userPermission);    
                    
                        pathAccessControlEntries.add(userEntry);

                    //      String continuationToken = null;

                    //    continuationToken = ResumeSetACLRecursively
                    //        (fileSystemClient, directoryClient, pathAccessControlEntries, continuationToken);

                        ContinueOnFailure(fileSystemClient, directoryClient, pathAccessControlEntries);

                    break;
                    case "9":
                        return;
                    default:
                        break;
                    }
                }
        } catch (java.lang.Exception e) {
            System.out.println(e.toString());
            System.exit(-1);

        }
    }
}
